### ISSUE TITLE
[desktop][windows] something wrong with windows desktop application destructor when it refers two or more plugins

### ISSUE THREAD

**Issue by yinquanyy** [flutter/flutter#93234]
Windows desktop applications refer to two or more plugins. When the application is closed, only one plugin C++ destructor is called

the first plugin's destructor
```
MyjdcPlugin::~MyjdcPlugin() {
  std::cout << "MyjdcPlugin::~MyjdcPlugin()" << std::endl;
}
```

the second plugin's destructor
```
GojdcPlugin::~GojdcPlugin() {
  std::cout << "GojdcPlugin::~GojdcPlugin()" << std::endl;
}
```

when the app closed, only output:
```
MyjdcPlugin::~MyjdcPlugin()
```

flutter --version
```
Flutter 2.5.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 18116933e7 (3 weeks ago) • 2021-10-15 10:46:35 -0700
Engine • revision d3ea636dc5
Tools • Dart 2.14.4
```

**Comment by darshankawar** [comment]
@yinquanyy 
Can you provide us steps to replicate and minimal code sample that shows this behavior ?
Thanks.

**Comment by yinquanyy** [comment]
> @yinquanyy Can you provide us steps to replicate and minimal code sample that shows this behavior ? Thanks.

ok

1. create project:
flutter create testjdc

mkdir plugins
cd plugins

2. create first plugin:
flutter create --template=plugin --platforms=windows gojdc
and then change the GojdcPlugin::~GojdcPlugin() code to
```
GojdcPlugin::~GojdcPlugin() {
  std::cout << "GojdcPlugin::~GojdcPlugin()" << std::endl;
}
```

3. create another plugin:
flutter create --template=plugin --platforms=windows myjdc
and then change the MyjdcPlugin::~MyjdcPlugin() code to
```
MyjdcPlugin::~MyjdcPlugin() {
  std::cout << "MyjdcPlugin::~MyjdcPlugin()" << std::endl;
}
```

4. add gojdc & myjdc to testjdc's pubspec.yaml
```
dependencies:
  flutter:
    sdk: flutter


  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^1.0.2
  myjdc:
    path: ./plugins/myjdc
  gojdc:
    path: ./plugins/gojdc
```

5. run the app:
flutter run testjdc

6. close the app. and console output:
```
MyjdcPlugin::~MyjdcPlugin()
```

**Comment by darshankawar** [comment]
Can you check [this](https://github.com/flutter/flutter/issues/64958) issue and see it could be the reason for the behavior you are seeing ?

**Comment by stuartmorgan** [comment]
Potential support for C# is unrelated to C++ plugin destruction.

This is an internal API mistake. Currently we have `FlutterDesktopPluginRegistrarSetDestructionHandler`, which I designed under the expectation of a singleton C++ plugin registrar (which is what uses that API). But because plugins are separate DLLs, I ran into ODR violations and had to change to having one private wrapper per plugin, and I forgot about the way this API was implemented. So what's happening is that each plugin is calling this, each time stomping the handler with their own.

We need to fix
https://github.com/flutter/engine/blob/2962099077b36704071802ea5595f4a016a1c214/shell/platform/windows/flutter_windows.cc#L177
to pass the registrar in as well, and keep a map of callbacks, not a single entry, in the engine. This should be a non-breaking change.
